CRM/Core/Form.php

```diff
@@ -364,11 +364,18 @@ public function registerRules() {
    */
   public function &add(
     $type, $name, $label = '',
-    $attributes = '', $required = FALSE, $extra = NULL
+    $attributes = NULL, $required = FALSE, $extra = NULL
   ) {
     if ($type === 'radio') {
       CRM_Core_Error::deprecatedFunctionWarning('CRM_Core_Form::addRadio');
     }
+
+    if ($attributes && !is_array($attributes)) {
+      // The $attributes param used to allow for strings and would default to an
+      // empty string.  However, now that the variable is heavily manipulated,
+      // we should expect it to always be an array.
```

Comment:
```suggestion
      // we should expect it to always be an array.
      $attributes = CRM_Core_BAO_CustomField::attributesFromString($attributes);
```

Reply:
@agh1 did you see this suggestion?

Reply:
Sorry, I didn't see it earlier.  My only concern is that at this point, string attributes coming in would be a surprise, probably only coming from an extension.  I don't believe there's anything I'm changing within this `add()` function that would cause further problems if the attributes were a string; it's just clear that much of the code doesn't anticipate that happening.

Assuming that if an extension does send a string of attributes, it must be working fine somehow, I'm not eager to introduce changes for that.  Logging the warning and letting the process continue should hopefully indicate to an extension developer that there's a problem without preemptively rearranging things for them.

Reply:
FYI one place seems to be [PDFLetterCommon](https://github.com/civicrm/civicrm-core/blob/194024315f33e507d3b11a295579749d255283c0/CRM/Core/Form/Task/PDFLetterCommon.php#L55). Comes up doing e.g. thank-you letters. I'll take a look.